### PR TITLE
JAMES-2890 Split JMAP RFC-8621 Serializer

### PR DESCRIPTION
**Pull request by chibenwa**
Rationals:

  - Many git conflicts in that class
  - This lengthy class is hard to read with almost no structure
  - Too much interleaving implicits...
  - Some static JSON deserialization requires a val serializer

 This is doable as we perform several parsing passes:

  - first the request structure
  - then back-reference substitutions
  - then method argument resolution
  - the */set creation and patches validation

These stages are independant from each other, and Methods are groupable by mostly distinct sets,
allowing a better modularisation.

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailQueryMethod.scala [hunk 1]
```diff
@@ -35,7 +35,7 @@ import org.reactivestreams.Publisher
 import play.api.libs.json.{JsError, JsSuccess}
 import reactor.core.scala.publisher.{SFlux, SMono}
 
-class EmailQueryMethod @Inject() (serializer: Serializer,
+class EmailQueryMethod @Inject() (serializer: EmailQuerySerializer,
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailQueryMethod.scala:38]: No need to inject this anymore, use `EmailQuerySerializer` directly line 68?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala [hunk 2]
```diff
@@ -43,7 +43,7 @@ import reactor.core.scheduler.Schedulers
 
 import scala.jdk.CollectionConverters._
 
-class MailboxGetMethod @Inject() (serializer: Serializer,
+class MailboxGetMethod @Inject() (serializer: MailboxSerializer,
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala:46]: Same remark

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala [hunk 3]
```diff
@@ -140,7 +140,7 @@ case class UpdateResults(results: Seq[UpdateResult]) {
   }).toMap
 }
 
-class MailboxSetMethod @Inject()(serializer: Serializer,
+class MailboxSetMethod @Inject()(serializer: MailboxSerializer,
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala:143]: Same remark

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailQueryMethod.scala (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailQueryMethod.scala:38]: The email query is still a variable.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala:46]: Idem, its not an object as there is a dependency to the MailboxId.Factory

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala (hunk 3 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxSetMethod.scala:143]: Idem, its not an object as there is a dependency to the MailboxId.Factory

**Review by Arsnael — APPROVED** [review]

**Review by remk — APPROVED** [review]

**Pull request merged**